### DISCUSSION TITLE
Question: Engine Development Tips

### DISCUSSION THREAD

**Discussion by jrruethe** [searx/searx#2572] [Development]
I'm interested in adding support for some additional engines and I'm really just looking for some development tips.

Right now, I am experimenting with the xpath engine. Because I am pretty new at this, I have been using the searx-docker container, so basically what I am doing is editing the xpath values in settings.yml, build and run the container, go to the webui and do a search, look at the logs to see what went wrong, then make an edit and try it all again.

This loop seems slow and inefficient to me, so I wanted to ask if there was a better way to do engine development. For example, is there a way to run a search using a single engine from the command line, perhaps in a way that I could get a debugger and inspect the variables? Is there a way to do a search one time and save the HTML, then have that same HTML file always returned so I can tinker with the xpath values without constantly hitting the external provider? I have looked through all the documentation I can find but I didn't see much along these lines.

Any engine development tips would be appreciated! Thank you

**Comment by unixfox** [comment]
Well you could launch Searx manually using python and enable the dev mode in the `settings.yaml`. In the debug mode every new change will reload Searx, so you can quickly test every new line of code without restarting Searx.

**Comment by dalf** [comment]
If you don't want to use the web interface, you can use the `utils/standalone_searx.py` script. 

See `python utils/standalone_searx.py --help`

You can also use this minimalist script:

```python
from searx.search import initialize, Search, SearchQuery, EngineRef

initialize()
search_query = SearchQuery('test', [EngineRef('your_new_engine_name', 'general')])
s = Search(search_query)
s.search()

for result in s.result_container.get_ordered_results():
    print(result)
```

**Reply by jrruethe** [reply]
Ah, Thank you! I was just returning to say that I figured out how to get searx running manually without Docker, and that is a bit faster. This command-line script appears to be exactly what I was looking for, I'll play with that.

Thanks again!

**Reply by dalf** [reply]
* I advise you to check using the web interface : `make run`  (or `source ./local/py3/bin/activate; SEARX_DEBUG=1 python searx/webapp.py` )
* also check there is no issue using `make test` (code formatting issue, unit test, etc...).

**Comment by jrruethe** [comment]
Thank you very much! I have been able to get going by working with that minimalist script above.
So far, I have been able to add engines for [Runaroo](https://runaroo.com) and [Million Short](https://millionshort.com).
I plan on doing a proper pull request for Runaroo later when I have it a bit more solid, but Million Short was able to be done with the standard json_engine:

```
      - name : millionshort
        engine : json_engine
        search_url : https://millionshort.com/api/search?keywords={query}&offset={pageno}&remove=1000000&shopping=y
        results_query : content/webPages
        url_query : displayUrl
        title_query : name
        content_query : snippet
        paging: True
        page_size: 10
        first_page_num: 0
        categories : general
        shortcut : ms
        disabled : False
```
